### PR TITLE
Improve test helpers

### PR DESCRIPTION
**Pull request by artem-lunarg**
Commit 1.  Reverts a3d60640e6f0cb403493d4744881b5e78a8955f8.
That change duplicates what `EXPECT(set_memory_type())` already does, and arguably `set_memory_type()` does it better because it checks only memory types supported by the resource but not all types. Successfully run local test session without this code.

Commit 2.  Adds optional `pNext` to create_info/allocate_info helpers which is similar to how `pNext` is used in other places (for example, `LvLInitStruct`).

Commit 3.  Allows to use `vk_testing::Buffer` and `vk_testing::Image` in the negative tests when BindMemory causes validation error.

Commit 4.  (thanks Juan!) Removes `DeviceMemory::alloc_info` which brings little value and instead `DeviceMemory::get_resource_alloc_info`  can provide  a larger chunk of functionality.

### CODE REVIEW

**Review by juan-lunarg — DISMISSED** [review]

**Comment by artem-lunarg** [comment]
Added two more commits here after @juan-lunarg approved the first commit. The new approval is needed :).

These changes allow to write more compact tests by allowing to use `vk_testing::Image` or `vk_testing::Buffer` in more scenarios in the negative tests. So far, these helper classes could be used in negative tests to test errors during resource creation. To be able to do that we allowed `VK_ERROR_VALIDATION_FAILED_EXT` in the resource creation routine:
```
#define NON_DISPATCHABLE_HANDLE_INIT(create_func, dev, ...)                                                 \
    do {                                                                                                    \
        assert(!initialized());                                                                             \
        handle_type handle;                                                                                 \
        auto result = create_func(dev.handle(), __VA_ARGS__, NULL, &handle);                                \
        if (EXPECT((result == VK_SUCCESS) || (result == VK_ERROR_VALIDATION_FAILED_EXT) ||                  \
...
```

This change allows `VK_ERROR_VALIDATION_FAILED_EXT` in BindMemory routines too.

**Comment by artem-lunarg** [comment]
Example of test simpification from ongoing PR. 

__Before:__
```
TEST_F(VkLayerTest, BufferMemoryWithUnsupportedExternalHandleType) {
    TEST_DESCRIPTION("Bind buffer memory with unsupported external memory handle type.");
    SetTargetApiVersion(VK_API_VERSION_1_1);
    ASSERT_NO_FATAL_FAILURE(InitFramework());
    if (DeviceValidationVersion() < VK_API_VERSION_1_1) {
        GTEST_SKIP() << "At least Vulkan version 1.1 is required";
    }
    ASSERT_NO_FATAL_FAILURE(InitState());

    // Create export buffer
    auto external_buffer_info = LvlInitStruct<VkExternalMemoryBufferCreateInfo>();
    auto buffer_info = LvlInitStruct<VkBufferCreateInfo>(&external_buffer_info);
    buffer_info.usage = VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT;
    buffer_info.size = 4096;

    const auto exportable_types =
        FindSupportedExternalMemoryHandleTypes(*this, buffer_info, VK_EXTERNAL_MEMORY_FEATURE_EXPORTABLE_BIT);
    if (!exportable_types) {
        GTEST_SKIP() << "Unable to find exportable handle type";
    }
    if (exportable_types == AllVkExternalMemoryHandleTypeFlagBits) {
        GTEST_SKIP() << "This test requires at least one unsupported handle type, but all handle types are supported";
    }
    const auto handle_type = LeastSignificantFlag<VkExternalMemoryHandleTypeFlagBits>(exportable_types);
    external_buffer_info.handleTypes = handle_type;

    VkBuffer buffer;
    if (vk::CreateBuffer(device(), &buffer_info, NULL, &buffer) != VK_SUCCESS) {
        GTEST_SKIP() << "Unable to create a valid external buffer";
    }

    // Create export memory with both supported and unsupported handle types
    const auto not_supported_type = LeastSignificantFlag<VkExternalMemoryHandleTypeFlagBits>(~exportable_types);
    auto export_memory_info = LvlInitStruct<VkExportMemoryAllocateInfo>();
    export_memory_info.handleTypes = handle_type | not_supported_type;

    auto alloc_info = LvlInitStruct<VkMemoryAllocateInfo>(&export_memory_info);
    VkMemoryRequirements mem_reqs;
    vk::GetBufferMemoryRequirements(device(), buffer, &mem_reqs);
    alloc_info.allocationSize = mem_reqs.size;
    DeviceObj()->phy().set_memory_type(mem_reqs.memoryTypeBits, &alloc_info, VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
    VkDeviceMemory memory;
    vk::AllocateMemory(device(), &alloc_info, nullptr, &memory);
    ASSERT_TRUE(memory != VK_NULL_HANDLE);

    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-VkExportMemoryAllocateInfo-handleTypes-00656");
    auto bind_info = LvlInitStruct<VkBindBufferMemoryInfo>();
    bind_info.buffer = buffer;
    bind_info.memory = memory;
    vk::BindBufferMemory2(device(), 1, &bind_info);
    m_errorMonitor->VerifyFound();
    vk::FreeMemory(device(), memory, nullptr);
    vk::DestroyBuffer(device(), buffer, nullptr);
}
```

__After:__
```
TEST_F(VkLayerTest, BufferMemoryWithUnsupportedExternalHandleType) {
    TEST_DESCRIPTION("Bind buffer memory with unsupported external memory handle type.");
    SetTargetApiVersion(VK_API_VERSION_1_1);
    ASSERT_NO_FATAL_FAILURE(InitFramework());
    if (DeviceValidationVersion() < VK_API_VERSION_1_1) {
        GTEST_SKIP() << "At least Vulkan version 1.1 is required";
    }
    ASSERT_NO_FATAL_FAILURE(InitState());

    auto external_buffer_info = LvlInitStruct<VkExternalMemoryBufferCreateInfo>();
    const auto buffer_info =
        vk_testing::Buffer::create_info(4096, VK_BUFFER_USAGE_TRANSFER_DST_BIT, nullptr, &external_buffer_info);
    const auto exportable_types =
        FindSupportedExternalMemoryHandleTypes(*this, buffer_info, VK_EXTERNAL_MEMORY_FEATURE_EXPORTABLE_BIT);
    if (!exportable_types) {
        GTEST_SKIP() << "Unable to find exportable handle type";
    }
    if (exportable_types == AllVkExternalMemoryHandleTypeFlagBits) {
        GTEST_SKIP() << "This test requires at least one unsupported handle type, but all handle types are supported";
    }
    const auto handle_type = LeastSignificantFlag<VkExternalMemoryHandleTypeFlagBits>(exportable_types);
    const auto not_supported_type = LeastSignificantFlag<VkExternalMemoryHandleTypeFlagBits>(~exportable_types);
    external_buffer_info.handleTypes = handle_type;
    auto export_memory_info = LvlInitStruct<VkExportMemoryAllocateInfo>();
    export_memory_info.handleTypes = handle_type | not_supported_type;

    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-VkExportMemoryAllocateInfo-handleTypes-00656");
    vk_testing::Buffer buffer(*m_device, buffer_info, VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT, &export_memory_info);
    m_errorMonitor->VerifyFound();
}
```

File: tests/vktestbinding.cpp [hunk 1]
```diff
@@ -489,16 +489,8 @@ void *DeviceMemory::map(VkFlags flags) {
...
-    VkPhysicalDeviceMemoryProperties dev_mem_props = dev.phy().memory_properties();
-    uint32_t mem_type_index = 0;
-    for (mem_type_index = 0; mem_type_index < dev_mem_props.memoryTypeCount; ++mem_type_index) {
-        if (mem_props == (mem_props & dev_mem_props.memoryTypes[mem_type_index].propertyFlags)) break;
-    }
-    // If we exceeded types, then this device doesn't have the memory we need
-    assert(mem_type_index < dev_mem_props.memoryTypeCount);
-    VkMemoryAllocateInfo info = alloc_info(reqs.size, mem_type_index);
+                                                           VkMemoryPropertyFlags mem_props, void *alloc_info_pnext) {
+    VkMemoryAllocateInfo info = alloc_info(reqs.size, 0, alloc_info_pnext);
```
Comment (juan-lunarg) [inline tests/vktestbinding.cpp:493]: If this has become more of a general cleanup of the tests. I'd remove alloc_info entirely. It provides little value and risks being used directly.

EX:
https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/59aa803960bd0b19e5369082a4b78eac16957498/tests/vklayertests_buffer_image_memory_sampler.cpp#L910

Comment (artem-lunarg) [reply tests/vktestbinding.cpp:493]: I agree, I was thinking at one moment about this too and now one more confirmation. This change is not strictly an attempt to clean up existing tests, mostly to prepare functionality that is needed in another PR, but yes, it's a good opportunity to make this change.

File: tests/vktestbinding.cpp [hunk 2]
```diff
@@ -736,7 +735,8 @@ VkMemoryRequirements Image::memory_requirements() const {
 }
 
 void Image::bind_memory(const DeviceMemory &mem, VkDeviceSize mem_offset) {
-    EXPECT(vk::BindImageMemory(device(), handle(), mem.handle(), mem_offset) == VK_SUCCESS);
+    const auto result = vk::BindImageMemory(device(), handle(), mem.handle(), mem_offset);
+    EXPECT(result == VK_SUCCESS || result == VK_ERROR_VALIDATION_FAILED_EXT);
```
Comment (juan-lunarg) [inline tests/vktestbinding.cpp:739]: What's the rational for `VK_ERROR_VALIDATION_FAILED_EXT`? Maybe worth a comment.

File: tests/vktestbinding.cpp [hunk 3]
```diff
@@ -670,7 +662,8 @@ VkMemoryRequirements Buffer::memory_requirements() const {
 }
 
 void Buffer::bind_memory(const DeviceMemory &mem, VkDeviceSize mem_offset) {
-    EXPECT(vk::BindBufferMemory(device(), handle(), mem.handle(), mem_offset) == VK_SUCCESS);
+    const auto result = vk::BindBufferMemory(device(), handle(), mem.handle(), mem_offset);
+    EXPECT(result == VK_SUCCESS || result == VK_ERROR_VALIDATION_FAILED_EXT);
```
Comment (juan-lunarg) [inline tests/vktestbinding.cpp:666]: What's the rational for `VK_ERROR_VALIDATION_FAILED_EXT`? Maybe worth a comment.

File: tests/vktestbinding.cpp (hunk 2 above)
Comment (artem-lunarg) [reply tests/vktestbinding.cpp:739]: It's the error generated by VVL when vulkan call causes validation error.
The reason is the same as in the `NON_DISPATCHABLE_HANDLE_INIT`. Only  EXPECTing `VK_SUCCESS` will result in test failure when validation error happens and `m_errorMonitor->VerifyFound()`, for example, won't be able to react. Here we allow either success calls or calls that fail due to validation (but not real vulkan failures).

**Review by juan-lunarg — APPROVED** [review]

**Comment by artem-lunarg** [comment]
Jenkins CI failure is not related. VkVideoLayerTest.InUseDestroyed.
